src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -665,9 +665,15 @@ export class SavedObjectsRepository {
 
     const deleteDocNotFound = body.result === 'not_found';
     const deleteIndexNotFound = body.error && body.error.type === 'index_not_found_exception';
+    const esServerSupported = isSupportedEsServer(headers);
     if (deleteDocNotFound || deleteIndexNotFound) {
-      // see "404s from missing index" above
-      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      if (esServerSupported) {
+        // see "404s from missing index" above
+        throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      } else {
+        // throw if we can't verify the response is from Elasticsearch
+        throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError(type, id);
```

Comment:
Self-review:
`createGenericNotFoundEsUnavailableError` will return a [503](https://github.com/elastic/kibana/pull/107104/files#diff-c6ac0127a77f2ba335589ef96b96203ca6b393d22a1878e95225c912c0e86d51R489) error that adds [the message from a `NotFoundError`](https://github.com/elastic/kibana/pull/107104/files#diff-c6ac0127a77f2ba335589ef96b96203ca6b393d22a1878e95225c912c0e86d51R483) to the `EsUnavailableError`. The decorated error that's thrown will [assert `true` to a check for `isEsUnavailableError`](https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/service/lib/errors.ts#L195), since the latter returns a `boolean` for the `statusCode` being 503.
If a saved object type and id are provided, the error will have the form:
```
{
	message: 'x-elastic-product not present or not recognized: Saved object [foo/bar] not 	found',
	statusCode: 503
}
```
Otherwise, the error will have the form:
```
{
	message: 'x-elastic-product not present or not recognized: Not Found',
	statusCode: 503
}
```

Reply:
optional nit: I had to read `createGenericNotFoundEsUnavailableError` implementation to understand what error it will create. Maybe we should pass an error reason to `decorateEsUnavailableError` explicitly?
```js
decorateEsUnavailableError(error, `Saved object [${type}/${id}] not found`);
```

> // throw if we can't verify the response is from Elasticsearch


Will we add such a helper for every possible error response?

Reply:
@joshdover which question did you give the thumbs up for in https://github.com/elastic/kibana/pull/107301#discussion_r682480245? i.e.

> Maybe we should pass an error reason to decorateEsUnavailableError explicitly?

or 

> Will we add such a helper for every possible error response?

I'm not sure which way we want to go eventually, as long as we provide enough info to easily figure out the root cause of an error when it's thrown.

Reply:
> Maybe we should pass an error reason to decorateEsUnavailableError explicitly?

We would still need to provide an error to pass to `decorateEsUnavailableError`:
```
public static decorateEsUnavailableError(error: Error, reason?: string) {
    return decorate(error, CODE_ES_UNAVAILABLE, 503, reason);
  }
```
which is what we're doing in `createGenericNotFoundEsUnavailableError`.